server/hotrod/src/main/java/org/infinispan/server/hotrod/tx/table/TxState.java

```diff
@@ -0,0 +1,163 @@
+package org.infinispan.server.hotrod.tx.table;
+
+import static org.infinispan.commons.marshall.MarshallUtil.marshallCollection;
+import static org.infinispan.commons.marshall.MarshallUtil.unmarshallCollection;
+import static org.infinispan.server.hotrod.tx.table.Status.ACTIVE;
+import static org.infinispan.server.hotrod.tx.table.Status.PREPARING;
+
+import java.io.IOException;
+import java.io.ObjectInput;
+import java.io.ObjectOutput;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+
+import org.infinispan.commands.write.WriteCommand;
+import org.infinispan.commons.marshall.AdvancedExternalizer;
+import org.infinispan.remoting.transport.Address;
+import org.infinispan.server.core.ExternalizerIds;
+import org.infinispan.transaction.xa.GlobalTransaction;
+import org.infinispan.util.TimeService;
+
+import net.jcip.annotations.Immutable;
+
+/**
+ * A transaction state stored globally in all the cluster members.
+ *
+ * @author Pedro Ruivo
+ * @since 9.4
+ */
+@Immutable
+public class TxState {
+
+   public static final AdvancedExternalizer<TxState> EXTERNALIZER = new Externalizer();
+   private final GlobalTransaction globalTransaction;
+   private final Status status;
+   private final List<WriteCommand> modifications;
+   private final boolean recoverable;
+   private final long timeout; //ms
+   private final long lastAccessTimeNs; //ns
```

Comment:
No need for the comment here, `ns` is already in the name.

But I'm not sure using nanos and `nanoTime()` is the right approach in a class that is sent remotely:

     * <p>The values returned by this method become meaningful only when
     * the difference between two such values, obtained within the same
     * instance of a Java virtual machine, is computed.

Reply:
it isn't sent remotely. the class is updated via functional API, so it uses the local TimeService and not the remote value.

Reply:
The externalizer does read+write the timestamp... maybe the externalizer isn't needed at all?

Reply:
I going to need to fix it later. The externalizer is used mainly by the state transfer and in that case, it is value is "garbage". In any case, it ignores the access time if the originator is still in the view :)